imod/mf6/utilities/clip.py

```diff
@@ -117,9 +117,14 @@ def clip_by_grid(package: ILineDataPackage, active: GridDataArray) -> ILineDataP
     bounding_gdf = bounding_polygon(active)
     package_gdf_clipped = package_gdf.clip(bounding_gdf)
     # Catch edge case: when line crosses only vertex of polygon, a point
-    # (type_id == 0) is returned. Drop these.
-    is_points = shapely.get_type_id(package_gdf_clipped.geometry) == 0
+    # or multipoint is returned. Drop these.
+    type_ids = shapely.get_type_id(package_gdf_clipped.geometry)
+    is_points = (type_ids == shapely.GeometryType.POINT) | (
+        type_ids == shapely.GeometryType.MULTIPOINT
+    )
     package_gdf_clipped = package_gdf_clipped[~is_points]
+    # Separate MultiLinestrings ino separate Linestrings
+    package_gdf_clipped = package_gdf_clipped.explode("geometry", ignore_index=True)
```

Comment:
This method is calling methods that are not on the ILineDataPackage interface but on the implementation of that interface. Therefor it is violating the Liskov substitution principle.  To avoid this i have altered this methods slightly:

```
@typedispatch
def clip_by_grid(package: ILineDataPackage, active: GridDataArray) -> ILineDataPackage:
    """Clip LineDataPackage outside unstructured/structured grid."""

    # Clip line with polygon
    bounding_gdf = bounding_polygon(active)
    clipped_geometry = package.geometry.clip(bounding_gdf)

    # Catch edge case: when line crosses only vertex of polygon, a point
    # or multipoint is returned. Drop these.
    type_ids = shapely.get_type_id(clipped_geometry.geometry)
    is_points = (type_ids == shapely.GeometryType.POINT) | (
        type_ids == shapely.GeometryType.MULTIPOINT
    )
    clipped_geometry = clipped_geometry[~is_points]

    # Convert MultiLineStrings to LineStrings
    clipped_geometry = clipped_geometry.transform({'geometry': lambda x: x if shapely.get_type_id(x) == 1 else shapely.ops.linemerge(x)})

    # Create new instance
    clipped_package = deepcopy(package)
    clipped_package.geometry = clipped_geometry
    return clipped_package
```

This has a benefit that the _getsettings, _get_variables_names_for_gdf and the _line_package_to_gdf methods are no longer needed and can be removed.


2 more changes are needed.  The ILineDataPackage becomes:
```
class ILineDataPackage(IPackageBase):
    """
    Interface for packages for which the data is defined by lines independent of the domain definition.
    """
    @property
    @abstractmethod
    def geometry(self) -> gpd.GeoDataFrame:
        raise NotImplementedError

    @geometry.setter
    @abstractmethod
    def geometry(self, value: gpd.GeoDataFrame) -> None:
        raise NotImplementedError
```

and the implementations for those methods in the hfb package are:
```
    @property
    def geometry(self) -> gpd.GeoDataFrame:
        return gpd.GeoDataFrame(geometry=self.dataset["geometry"].values)

    @geometry.setter
    def geometry(self, value: gpd.GeoDataFrame) -> None:
        self.dataset["geometry"] = value.to_xarray()["geometry"]
        self.dataset = self.dataset.dropna(dim="index")
```